RFC for supporting mixed case identifiers (#36)
prestodb/rfcs: RFC-0010-mixed-case-identifier-support.md

# **RFC-0010 for Presto**

## Mixed case identifiers

Proposers

* Reetika Agrawal

## Summary

Improve Presto's identifier (schema, table & column names) handling to align with SQL standards, ensuring better interoperability with case-sensitive
and case-normalizing databases while minimizing SPI-breaking changes.

## Background

> Presto treats all identifiers as case-insensitive, normalizing them (typically to lowercase). This creates issues when

Reviewer: ```suggestion
Presto treats all identifiers as case-insensitive, normalizing them to lowercase. This creates issues when
```

querying databases that are case-sensitive (e.g., MySQL, PostgreSQL) or case-normalizing to uppercase (e.g., Oracle,
DB2). Without a standard approach, identifiers might not match the actual names in the underlying data sources, leading
to unexpected query failures or incorrect results. Additionally, inconsistent handling of delimited and non-delimited

> identifiers across different connectors further complicates cross-engine compatibility.

Reviewer: Can you elaborate more on `inconsistent handling of delimited and non-delimited
identifiers`

> The goal here is to improve interoperability with storage engines by aligning identifier handling with SQL standards

Reviewer: Goals can be listed as bullet points:

* Improve interoperability with storage engines i.e standardizing identifier handling (SQL Standard ref: [ANSI?](https://www.ibm.com/docs/en/netezza?topic=md-handle-sql-identifiers#c_dbuser_handle_sql_identifiers__title__2)).
* Backwards compatibility with existing system. Minimal or no breaking changes to SPI.

*

while ensuring a seamless user experience. Ideally, the change should be implemented in a way that minimizes

> backward-compatibility-breaking changes to the SPI, allowing connectors to adopt the new approach without significant

Reviewer: ```suggestion
breaking changes to the SPI, i.e. allowing connectors to adopt the new approach without significant
```

modifications as per their own requirements.

### Proposed Plan

Connectors handle identifiers in three ways:

1. SQL-Compliant (e.g., Oracle, DB2)
  - Delimited identifiers keep their original case.
  - Non-delimited identifiers are converted to uppercase.

2. Case-Sensitive (e.g., PostgreSQL)
  - Delimited identifiers keep their original case.
  - Non-delimited identifiers may be converted to lowercase or another case.

3. Case-Insensitive (e.g., Hive)
  - Delimited and non-delimited identifiers are treated the same.
  - Identifiers may be automatically converted to a specific case.

Presto handles identifiers in several ways:

  - Matching identifiers to locate entities such as catalogs, schemas, tables, views.
  - Resolving column names based on table metadata provided by connectors.
  - Passing identifiers to connectors when creating new entities.
  - Processing and displaying entity names retrieved from connectors, including column resolution and metadata introspection commands like SHOW and DESCRIBE.

## Proposed Implementation

#### Core Changes

> * In the common code path, make changes to pass original identifier (Schema, table and column names)

Reviewer: ```suggestion
* In the presto-spi, make changes to pass original identifier (Schema, table and column names)
```

#### MySQL Table Handling

```
presto> show schemas from mysql;
       Schema       
--------------------
 Test               
 TestDb             
 information_schema 
 performance_schema 
 sys                
 testdb             
(6 rows)

presto> show tables from mysql.TestDb;
   Table   
-----------
 Test      
 TestTable 
 testtable 
(3 rows)

presto> SHOW CREATE TABLE mysql.TestDb.Test;
             Create Table             
--------------------------------------
 CREATE TABLE mysql."TestDb"."Test" ( 
    "id" integer,                     
    "Name" char(10)                   
 )                                    
(1 row)

presto> select * from mysql.TestDb.Test;
 id |    Name    
----+------------
  2 | Tom        
(1 row)
```

## Backward Compatibility Considerations

* Existing connectors that do not implement normalizeIdentifier will default to lowercase normalization.
* Any connectors requiring case preservation can override the default behavior.
* A configuration flag could be introduced to allow backward-compatible identifier handling at the catalog level.

## Test Plan

* Ensure that existing CI tests pass for connectors where no specific implementation is added.

> * Add support for mixed-case identifiers in at least one JDBC connector (e.g., MySQL, PostgreSQL) and create relevant unit tests.

Reviewer: ```suggestion
* Add unit tests for testing mixed-case identifiers support in a JDBC connector (e.g., MySQL, PostgreSQL).
```
